src/Skia/Avalonia.Skia/DrawingContextImpl.cs

```diff
@@ -184,19 +184,116 @@ public void DrawGeometry(IBrush brush, IPen pen, IGeometryImpl geometry)
+            public static BoxShadowFilter Create(BoxShadow shadow, double opacity, bool skipDilate)
+            {
+                var ac = shadow.Color;
+                var spread = (int)shadow.Spread;
+                if (shadow.IsInset)
+                    spread = -spread;
+                
+                var filter = SKImageFilter.CreateDropShadow(
+                    (float)shadow.OffsetX,
+                    (float)shadow.OffsetY,
+                    (float)shadow.Blur / 2,
+                    (float)shadow.Blur / 2,
+                    new SKColor(ac.R, ac.G, ac.B, (byte)(ac.A * opacity)),
+                    SKDropShadowImageFilterShadowMode.DrawShadowOnly, null);
+                
+                var paint = new SKPaint { Color = SKColors.White, ImageFilter = filter };
+                
+                return new BoxShadowFilter
+                {
+                    Paint = paint, _filter = filter,
+                    ClipOperation = shadow.IsInset ? SKClipOperation.Intersect : SKClipOperation.Difference
+                };
+            }
+
+            public void Dispose()
+            {
+                Paint.Dispose();
+                _filter.Dispose();
+            }
+        }
+
+        SKRect AreaCastingShadowInHole(
+            SKRect hole_rect,
+            float shadow_blur,
+            float shadow_spread,
+            float offsetX, float offsetY)
+        {
+            // Adapted from Chromium
+            var bounds = hole_rect;
+
+            bounds.Inflate(shadow_blur, shadow_blur);
+
+            if (shadow_spread < 0)
+                bounds.Inflate(-shadow_spread, -shadow_spread);
+
+            var offset_bounds = bounds;
+            offset_bounds.Offset(-offsetX, -offsetY);
+            bounds.Union(offset_bounds);
+            return bounds;
+        }
+
+
         /// <inheritdoc />
-        public void DrawRectangle(IBrush brush, IPen pen, Rect rect, double radiusX, double radiusY)
+        public void DrawRectangle(IBrush brush, IPen pen, RoundedRect rect, BoxShadow boxShadow = default)
         {
-            var rc = rect.ToSKRect();
-            var isRounded = Math.Abs(radiusX) > double.Epsilon || Math.Abs(radiusY) > double.Epsilon;
+            var rc = rect.Rect.ToSKRect();
+            var isRounded = rect.IsRounded;
+            var needRoundRect = rect.IsRounded || (!boxShadow.IsEmpty && boxShadow.IsInset);
+            using var skRoundRect = needRoundRect ? new SKRoundRect() : null;
+            if (needRoundRect)
+                skRoundRect.SetRectRadii(rc,
+                    new[]
+                    {
+                        rect.RadiiTopLeft.ToSKPoint(), rect.RadiiTopRight.ToSKPoint(),
+                        rect.RadiiBottomRight.ToSKPoint(), rect.RadiiBottomLeft.ToSKPoint(),
+                    });
 
+            if (!boxShadow.IsEmpty && !boxShadow.IsInset)
+            {
+                using(var shadow = BoxShadowFilter.Create(boxShadow, _currentOpacity, true))
```

Comment:
Can this be shared by stroke and fill operation?

Reply:
There are two different filters for inset and normal shadows. Shadow is drawn even if there is no fill brush and stroke pen.